ring/lifecycler_test.go

```diff
@@ -98,6 +98,113 @@ func TestLifecycler_HealthyInstancesCount(t *testing.T) {
+	t.Cleanup(func() { assert.NoError(t, closer.Close()) })
+
+	var ringConfig Config
+	flagext.DefaultValues(&ringConfig)
+	ringConfig.KVStore.Mock = ringStore
+
+	instances := []struct {
+		zone                                string
+		healthy                             bool
+		expectedHealthyInstancesInZoneCount int
+		expectedHealthyInstancesCount       int
+	}{
+		{
+			zone:    "zone-a",
+			healthy: true,
+			// after adding a healthy instance in zone-a, expectedHealthyInstancesInZoneCount in zone-a becomes 1
+			expectedHealthyInstancesInZoneCount: 1,
+			// after adding a healthy instance in zone-a, expectedHealthyInstancesCount is 1
+			expectedHealthyInstancesCount: 1,
+		},
+		{
+			zone:    "zone-a",
+			healthy: false,
+			// after adding an unhealthy instance in zone-a, expectedHealthyInstancesInZoneCount in zone-a remains 1
+			expectedHealthyInstancesInZoneCount: 1,
+			// zone-a was already added, so expectedHealthyInstancesCount remains 1
+			expectedHealthyInstancesCount: 1,
+		},
+		{
+			zone:    "zone-a",
+			healthy: true,
+			// after adding a healthy instance in zone-a, expectedHealthyInstancesInZoneCount in zone-a becomes 2
+			expectedHealthyInstancesInZoneCount: 2,
+			// zone-a was already added, so expectedHealthyInstancesCount remains 1
+			expectedHealthyInstancesCount: 1,
+		},
+		{
+			zone:    "zone-b",
+			healthy: true,
+			// after adding a healthy instance in zone-b, expectedHealthyInstancesInZoneCount in zone-b becomes 1
+			expectedHealthyInstancesInZoneCount: 1,
+			// after adding a healthy instance in zone-b, expectedHealthyInstancesCount becomes 2
+			expectedHealthyInstancesCount: 2,
+		},
+		{
+			zone:    "zone-c",
+			healthy: false,
+			// after adding an unhealthy instance in zone-c, expectedHealthyInstancesInZoneCount in zone-c remains 0
+			expectedHealthyInstancesInZoneCount: 0,
+			// after adding an unhealthy instance in zone-c, expectedHealthyInstancesCount becomes 3
+			expectedHealthyInstancesCount: 3,
+		},
+		{
+			zone:    "zone-c",
+			healthy: true,
+			// after adding a healthy instance in zone-c, expectedHealthyInstancesInZoneCount in zone-c becomes 1
+			expectedHealthyInstancesInZoneCount: 1,
+			// zone-c was already added, so expectedHealthyInstancesCount remains 3
+			expectedHealthyInstancesCount: 3,
+		},
+		{
+			zone:    "zone-b",
+			healthy: true,
+			// after adding a healthy instance in zone-b, expectedHealthyInstancesInZoneCount in zone-b becomes 2
+			expectedHealthyInstancesInZoneCount: 2,
+			// zone-b was already added, so expectedHealthyInstancesCount remains 3
+			expectedHealthyInstancesCount: 3,
+		},
+	}
+
+	expectedHealthInstancesCounter := 0
+	for idx, instance := range instances {
+		ctx := context.Background()
+
+		// Register an instance to the ring.
+		cfg := testLifecyclerConfig(ringConfig, fmt.Sprintf("instance-%d", idx))
+		cfg.HeartbeatPeriod = 100 * time.Millisecond
+		joinTimeoutMs := 1000
```

Comment:
[nit] It's not really a timeout, in the sense the instance will actually join after `joinAfterMs` but were not waiting for that in the test because this value is lower than 2 seconds.

```suggestion
		joinWaitMs := 1000
```